File: angle_calc.c
#include <stdio.h>
#include <math.h>
#include <stdlib.h>

#define deg_to_rad(angleInDegrees) ((angleInDegrees) * M_PI / 180.0)
#define rad_to_deg(angleInRadians) ((angleInRadians) * 180.0 / M_PI)

// The WGS85 ellipsoid semi-major axis and flattening
float WGS85_axis = 6378137;
float WGS85_flattening = 298.257223563;


/**
 * @param lat latitude in decimal degrees
 * @param lon longitude in decimal degrees
 * @param alt altitude in meters
 * @return a float array containing the X, Y and Z coordinates
 */
float* geodetic_to_geocentric(float lat, float lon, float alt){
	float *coordinates = (float*)malloc(sizeof(float) * 3);
	float lat_r = deg_to_rad(lat);
	float lon_r = deg_to_rad(lon);

	float e2 = 1 - pow((1 - 1 / WGS85_flattening), 2);
	float n = WGS85_axis / sqrt(1 - e2 * pow(sin(lat_r), 2));
	float r = (n + alt) * cos(lat_r);

	coordinates[0] = r * cos(lon_r);
	coordinates[1] = r * sin(lon_r);
	coordinates[2] = (n * (1 - e2) + alt) * sin(lat_r);

	return coordinates;
}


/**
 * @param source float array containing the X, Y and Z coordinates of the observer
 * @param destination lfloat array containing the X, Y and Z coordinates of the observable
 * @return the angle of elevation in decimal degrees
 */
float get_elevation(float* source, float* destination){
	float x = source[0];
	float y = source[1];
	float z = source[2];
	float dx = destination[0] - x;
	float dy = destination[1] - y;
	float dz = destination[2] - z;

	float cos_elevation = (x*dx + y*dy + z*dz) / 
						sqrt((pow(x, 2) + pow(y, 2) + pow(z, 2))*(pow(dx, 2) + pow(dy, 2) + pow(dz, 2)));
	float angle = rad_to_deg(acos(cos_elevation));

	return  90 - angle;
}


/**
 * @param source float array containing the X, Y and Z coordinates of the observer
 * @param destination lfloat array containing the X, Y and Z coordinates of the observable
 * @return the azimuth angle in decimal degrees
 */
float get_azimuth(float* source, float* destination){
	float x = source[0];
	float y = source[1];
	float z = source[2];
	float dx = destination[0] - x;
	float dy = destination[1] - y;
	float dz = destination[2] - z;

	float cos_azimuth = (-z*x*dx - z*y*dy + (pow(x, 2) + pow(y, 2))*dz) / 
				sqrt((pow(x, 2) + pow(y, 2)) * (pow(x, 2) + pow(y, 2) + pow(z, 2)) * (pow(dx, 2) + pow(dy, 2) + pow(dz, 2)));
	float sin_azimuth = (-y*dx + x*dy) / sqrt((pow(x, 2) + pow(y, 2)) * (pow(dx, 2) + pow(dy, 2) + pow(dz, 2)));

	float angle = rad_to_deg(atan2(sin_azimuth, cos_azimuth));

	return fmod((360 + angle), (float)360);
}

int main(int argc, char** argv){
	if(argc != 7){
		printf("Usage: %s [source_lat] [source_lon] [source_alt] [dest_lat] [dest_lon] [dest_alt]\n", argv[0]);
		exit(1);
	}

	float *source = geodetic_to_geocentric(strtof(argv[1], NULL), strtof(argv[2], NULL), strtof(argv[3], NULL));
	float *destination = geodetic_to_geocentric(strtof(argv[4], NULL), strtof(argv[5], NULL), strtof(argv[6], NULL));

	float elevation = get_elevation(source, destination);
	float azimuth = get_azimuth(source, destination);

	printf("Azimuth: %f degrees\n", azimuth);
	printf("Elevation: %f degrees\n", elevation);

	free(source);
	free(destination);
	return 0;
}

File: angle_calc.py
from math import cos, radians, sin, sqrt, acos, degrees, atan2
import sys

# Ellipsoid parameters: semi-major axis and flattening
WGS84 = 6378137, 298.257223563


def geodetic_to_geocentric(latitude, longitude, altitude):
	"""
	:param latitude: Latitude in decimal degrees
	:param longitude: Longitude in decimal degrees
	:param altitude: Altitude in meters above the ellipsoid
	:returns: The X, Y, Z geocentric coordinates
	"""
	lat_r = radians(latitude)
	lon_r = radians(longitude)
	sin_lat = sin(lat_r)
	a, rf = WGS84
	e2 = 1 - (1 - 1 / rf) ** 2
	n = a / sqrt(1 - e2 * sin_lat ** 2)
	r = (n + altitude) * cos(lat_r)
	x = r * cos(lon_r)
	y = r * sin(lon_r)
	z = (n * (1 - e2) + altitude) * sin_lat

	return x, y, z


def get_elevation(source, destination):
	"""
	:param source: Tuple with the x, y, z coordinates of the observer
	:param destination: Tuple with the x, y, z coordinates of the observable
	:returns: The elevation angle in decimal degrees of the observable
	"""
	x = source[0]
	y = source[1]
	z = source[2]
	dx = destination[0] - x
	dy = destination[1] - y
	dz = destination[2] - z

	cos_elevation = (x*dx + y*dy + z*dz) / sqrt((x**2 + y**2 + z**2)*(dx**2 + dy**2 + dz ** 2))
	angle = degrees(acos(cos_elevation))

	return 90 - angle


def get_azimuth(source, destination):
	"""
	:param source: Tuple with the x, y, z coordinates of the observer
	:param destination: Tuple with the x, y, z coordinates of the observable
	:returns: The azimuth angle in decimal degrees of the observable
	"""
	x = source[0]
	y = source[1]
	z = source[2]
	dx = destination[0] - x
	dy = destination[1] - y
	dz = destination[2] - z

	cos_azimuth = (-z*x*dx - z*y*dy + (x**2 + y**2)*dz) / \
					sqrt((x**2 + y**2) * (x**2 + y**2 + z**2) * (dx**2 + dy**2 + dz ** 2))
	sin_azimuth = (-y*dx + x*dy) / sqrt((x**2 + y**2) * (dx**2 + dy**2 + dz**2))

	angle = degrees(atan2(sin_azimuth, cos_azimuth))

	# if(angle < 0):
	# 	angle = 360 + angle
	# return angle
	return (360 + angle) % 360


if __name__ == "__main__":
	if(len(sys.argv) != 7):
		print("Usage: {} [source_lat] [source_lon] [source_alt] [dest_lat] [dest_lon] [dest_alt]".format(sys.argv[0]))
		sys.exit(1)

	source = geodetic_to_geocentric(float(sys.argv[1]), float(sys.argv[2]), float(sys.argv[3]))
	destination = geodetic_to_geocentric(float(sys.argv[4]), float(sys.argv[5]), float(sys.argv[6]))
	print("Azimuth: {} degrees".format(get_azimuth(source, destination)))
	print("Elevation: {} degrees".format(get_elevation(source, destination)))
